academic-observatory-workflows/academic_observatory_workflows/openalex_telescope/tasks.py

```diff
@@ -0,0 +1,683 @@
+        bash_command="mkdir -p "
+        + output_folder
+        + " && gcloud auth activate-service-account --key-file=${GOOGLE_APPLICATION_CREDENTIALS}"
+        + f" && gsutil -m -q cp -L {entity.log_path} -r "
+        + bucket_path
+        + " "
+        + output_folder,
+        do_xcom_push=False,
+    )
+    op.execute(context)
+
+
+def transform(*, entity: OpenAlexEntity):
+    # Cleanup in case we re-run task
+    output_folder = os.path.join(entity.transform_folder, "data", entity.entity_name)
+    if os.path.exists(output_folder):
+        clean_dir(output_folder)
+
+    # Initialise schema generator
+    merged_schema_map = OrderedDict()
+
+    with ProcessPoolExecutor(max_workers=os.cpu_count()) as executor:
+        futures = []
+        for entry in entity.entries:
+            input_path = os.path.join(entity.download_folder, entry.object_key)
+            output_path = os.path.join(entity.transform_folder, entry.object_key)
+            futures.append(executor.submit(transform_file, input_path, output_path))
+        for future in as_completed(futures):
+            input_path, schema_map, schema_error = future.result()
+
+            if schema_error:
+                logging.info(f"Error generating schema for file {input_path}: {schema_error}")
+
+            # Merge the schemas from each process. Each data file could have more fields than others.
+            merged_schema_map = merge_schema_maps(to_add=schema_map, old=merged_schema_map)
+
+    # Flatten schema from nested OrderedDicts to a regular Bigquery schema.
+    merged_schema = flatten_schema(schema_map=merged_schema_map)
+
+    # Save schema to file
+    with open(entity.generated_schema_path, mode="w") as f_out:
+        json.dump(merged_schema, f_out, indent=2)
+
+
+def upload_schema(*, entity: OpenAlexEntity, transform_bucket: str):
+    success = gcs_upload_files(bucket_name=transform_bucket, file_paths=[entity.generated_schema_path])
+    if not success:
+        raise AirflowException("upload_schema: error uploading schema")
+
+
+def compare_schemas(*, entity: OpenAlexEntity, transform_bucket: str, slack_conn_id: str, **context):
+    logging.info(f"Loading schemas from file: generated: {entity.generated_schema_path}")
+    logging.info(f"Expected: {entity.schema_file_path}")
+
+    # Download merged schema
+    success = gcs_download_blob(
+        bucket_name=transform_bucket,
+        blob_name=gcs_blob_name_from_path(entity.generated_schema_path),
+        file_path=entity.generated_schema_path,
+    )
+    if not success:
+        raise AirflowException("compare_schemas: error downloading schema")
+
+    # Read in the expected schema for the entity.
+    merged_schema = load_json(entity.generated_schema_path)
+    expected_schema = load_json(entity.schema_file_path)
+
+    try:
+        match = bq_compare_schemas(expected_schema, merged_schema, check_types_match=False)
+    except:
+        match = False
+
+    if not match:
```

Comment:
Will this eventually lead to an error?

Reply:
It is just for informational purposes, it compares the current schema to the generated schema based on the data that it loaded.

If new fields are added the data will still load, as we ignore new fields.

If there is a breaking change and the table doesn't load, then we can use the generated schema to help figure out why.